### PR TITLE
[StructuralMechanicsApp] Minor inconsistencies in rotations in Init and Fin solution step methods

### PR DESCRIPTION
**Pull request by AlejandroCornejo**
**📝 Description**
Together with @lagoncalvesjr we noticed that, for cases in which we rotate the local axis of the elements a few rotations when missing when doing the `InitializeSolutionStep`, `FinalizeSolutionStep` and `CalculateOnConstitutiveLaw` methods.

I've solved this issues together with removing redundant checks in the `InitializeSolutionStep` and in the `FinalizeSolutionStep`.

### CODE REVIEW

File: applications/StructuralMechanicsApplication/custom_elements/base_solid_element.h
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_elements/base_solid_element.h:1058]:
```suggestion
            // Rotate to local axes strain/F
```

Should we check first that the element is rotated to avoid overhead?

File: applications/StructuralMechanicsApplication/custom_elements/base_solid_element.cpp
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_elements/base_solid_element.cpp]:
```suggestion
            // Rotate to local axes strain/F
```

Iden, no check?

File: applications/StructuralMechanicsApplication/custom_elements/base_solid_element.cpp
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_elements/base_solid_element.cpp]:
```suggestion
            // Rotate to local axes strain/F
```

Idem ch

**Review by loumalouomega — CHANGES_REQUESTED** [review]
Comments

File: applications/StructuralMechanicsApplication/custom_elements/base_solid_element.h
Comment (AlejandroCornejo) [reply applications/StructuralMechanicsApplication/custom_elements/base_solid_element.h:1058]: At that time we did it like this to try to encapsulate... I can try to separate the check

Comment (AlejandroCornejo) [reply applications/StructuralMechanicsApplication/custom_elements/base_solid_element.h:1058]: Ideally this check should be done once per simulation but IDK when to do it... In the Initialize we may have the problem that the element has not been assigned the rotation.

Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/custom_elements/base_solid_element.h:1058]: This is important, otherwise could have a significative impact

**Comment by AlejandroCornejo** [comment]
okkey @loumalouomega , this took longer than expected. 

In order to do the check only once per element (before was once per IP) I had to de-encapsulate the rotations to be exposed in the upper method. It's less clean but more efficient. 

Next I had to to minor modifications to the solid elements (we lost some encapsulation).
The tests are passing in my computer. 

Additionally i found some inconsistencies when printing stresses (rotations were missing).

**Review by loumalouomega — COMMENTED** [review]
Now I agree, I will unblock if @RiccardoRossi agrees

**Comment by AlejandroCornejo** [comment]
> Now I agree, I will unblock if @RiccardoRossi agrees

ok let's wait for @RiccardoRossi

File: applications/StructuralMechanicsApplication/custom_elements/base_solid_element.cpp
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/custom_elements/base_solid_element.cpp]: you can have different CLs within the element...i think that the previosu code was better here

File: applications/StructuralMechanicsApplication/custom_elements/base_solid_element.cpp
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/custom_elements/base_solid_element.cpp:131]: what are u trying to do here?

File: applications/StructuralMechanicsApplication/custom_elements/base_solid_element.cpp
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/custom_elements/base_solid_element.cpp]: same, here previous code is better

File: applications/StructuralMechanicsApplication/custom_elements/base_solid_element.cpp
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/custom_elements/base_solid_element.cpp]: ?? what is this

**Review by RiccardoRossi — COMMENTED** [review]
this looks a pretty big change to me.

why do you need to cal lthe "SetConstitutiveVariables" so many times now?

File: applications/StructuralMechanicsApplication/custom_elements/base_solid_element.cpp
Comment (AlejandroCornejo) [reply applications/StructuralMechanicsApplication/custom_elements/base_solid_element.cpp]: ok! i did not think that this could happen

File: applications/StructuralMechanicsApplication/custom_elements/base_solid_element.cpp
Comment (AlejandroCornejo) [reply applications/StructuralMechanicsApplication/custom_elements/base_solid_element.cpp:131]: the main issue is that the check `IsElementRotated()` was done every time the `CalculateConstitutiveVariables` (so at each IP) and that was more expensive than putting the check outside the IP loop. I have an idea to encapsulate this better

File: applications/StructuralMechanicsApplication/custom_elements/base_solid_element.cpp
Comment (AlejandroCornejo) [reply applications/StructuralMechanicsApplication/custom_elements/base_solid_element.cpp]: done

File: applications/StructuralMechanicsApplication/custom_elements/base_solid_element.cpp
Comment (AlejandroCornejo) [reply applications/StructuralMechanicsApplication/custom_elements/base_solid_element.cpp]: now i've added a bool to the `CalculateConstitutiveVariables` in order to know if it's rotated. Now they are back as it was

**Comment by AlejandroCornejo** [comment]
With the latests changes the `CalculateConstitutiveVariables` method remained as it was but adding a const bool to know if it's rotated or not.

**Comment by AlejandroCornejo** [comment]
Besides I corrected the `FinalizeSolutionStep` and `InitializeSolutionStep` cause they were inconsistent (no ratations were performed)

**Comment by AlejandroCornejo** [comment]
ping

File: applications/StructuralMechanicsApplication/custom_elements/base_solid_element.cpp
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/custom_elements/base_solid_element.cpp:843]: but then, don't you need to change all of the derived elements?

File: applications/StructuralMechanicsApplication/custom_elements/base_solid_element.h
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/custom_elements/base_solid_element.h:777]: this function is virtual, so making this change will imply changing all of the derived elements

File: applications/StructuralMechanicsApplication/custom_elements/small_displacement_bbar.cpp
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/custom_elements/small_displacement_bbar.cpp:116]: just to tell (not really a review here, just to let you know) that you can use  the .clear()) method to do the same shorter

**Review by RiccardoRossi — APPROVED** [review]
overall ok to me. 

I am more and more convinced that our derivation strategy is too long here, and that we would be better off with standalone functions instead of implementing everything in the base class, but this is definitely off topic.

approving on my side

File: applications/StructuralMechanicsApplication/custom_elements/base_solid_element.cpp
Comment (AlejandroCornejo) [reply applications/StructuralMechanicsApplication/custom_elements/base_solid_element.cpp:843]: I did it, i changed the UL, TL and small displ elements :)

**Comment by AlejandroCornejo** [comment]
@loumalouomega

**Comment by loumalouomega** [comment]
I am out until Monday, does @RiccardoRossi agrees?

**Comment by AlejandroCornejo** [comment]
> I am out until Monday, does @RiccardoRossi agree?

he approved the PR

**Comment by loumalouomega** [comment]
Okay from my side, feel free to merge

**Comment by AlejandroCornejo** [comment]
@loumalouomega you should remove the block :)

**Review by loumalouomega — APPROVED** [review]
Approving after approval from @RiccardoRossi

**Pull request merged**
